### PR TITLE
oshmem: scoll: fixes strided alltoall

### PR DESCRIPTION
**Pull request by alex-mikheev**
Signed-off-by: Alex Mikheev <alexm@mellanox.com>

@yosefe @jladd-mlnx please take a look

### CODE REVIEW

File: oshmem/mca/scoll/basic/scoll_basic_alltoall.c [hunk 1]
```diff
@@ -96,28 +126,47 @@ static int _algorithm_simple(struct oshmem_group_t *group,
...
-    return rc;
+    SCOLL_VERBOSE(14,
+                  "[#%d] send data to all PE in the group",
+                  group->my_pe);
+    j = oshmem_proc_group_find_id(group, group->my_pe);
+    for (i = 0; i < group->proc_count; i++) {
+        /* index permutation for better distribution of traffic */
+        k = (((j)+(i))%(group->proc_count));
+        pe_cur = oshmem_proc_pe(group->proc_array[k]);
+        rc = MCA_SPML_CALL(put((char *)target + j * nelems * element_size,
```
Comment (yosefe) [inline oshmem/mca/scoll/basic/scoll_basic_alltoall.c]: maybe we can reuse and call `get_stride_elem(target, 1, nelems, element_size, j, 0)`

File: oshmem/mca/scoll/basic/scoll_basic_alltoall.c [hunk 2]
```diff
@@ -96,28 +126,47 @@ static int _algorithm_simple(struct oshmem_group_t *group,
...
         k = (((j)+(i))%(group->proc_count));
         pe_cur = oshmem_proc_pe(group->proc_array[k]);
-        rc = MCA_SPML_CALL(put(
-                (void *)((char *)target + j * tst * nelems * element_size),
-                nelems * element_size,
-                (void *)((char *)source + i * sst * nelems * element_size),
-                pe_cur));
-        if (OSHMEM_SUCCESS != rc) {
-            break;
+        for (l = 0; l < nelems; l++) {
```
Comment (yosefe) [inline oshmem/mca/scoll/basic/scoll_basic_alltoall.c]: pls rename (i,j,k,l) to meaningful names

File: oshmem/mca/scoll/basic/scoll_basic_alltoall.c [hunk 3]
```diff
@@ -36,57 +42,81 @@ int mca_scoll_basic_alltoall(struct oshmem_group_t *group,
...
+    }
+
+    /* fence (which currently acts as quiet) is needed
+     * because scoll level barrier does not guarantee put completion
+     */
+    MCA_SPML_CALL(fence());
+
+    /* Wait for operation completion */
+    SCOLL_VERBOSE(14, "[#%d] Wait for operation completion", group->my_pe);
+    rc = BARRIER_FUNC(group, (pSync + 1), SCOLL_DEFAULT_ALG);
```
Comment (yosefe) [inline oshmem/mca/scoll/basic/scoll_basic_alltoall.c]: no need for ( ) around `pSync + 1`

File: oshmem/mca/scoll/basic/scoll_basic_alltoall.c (hunk 1 above)
Comment (yosefe) [reply oshmem/mca/scoll/basic/scoll_basic_alltoall.c]: also seems like the 1st argument to put() can be pre-calculated before the loop

**Comment by yosefe** [comment]
@alex-mikheev pls squash and add signed off

**Comment by yosefe** [comment]
@alex-mikheev pls pick for v3.1.x

**Comment by sam6258** [comment]
Has this been picked into v3.1.x and v3.0.x? I'd be happy to make those PR's if not.

**Pull request merged**